File: lib/soundsync_web/channels/signaling_channel.ex
defmodule SoundsyncWeb.SignalingChannel do
  @moduledoc """
  WebRTC signaling channel: relays offer / answer / ICE between peers.
  """

  use Phoenix.Channel

  alias SoundsyncWeb.ProjectTopic

  require Logger

  @impl true
  def join("signaling:" <> workspace_id, _params, socket) do
    case ProjectTopic.authorize(workspace_id, socket.assigns.current_user) do
      {:ok, project} ->
        Logger.info("[Signaling:#{workspace_id}] peer joined")

        {:ok,
         socket
         |> assign(:workspace_id, workspace_id)
         |> assign(:project_id, project.id)}

      {:error, reason} ->
        {:error, %{reason: to_string(reason)}}
    end
  end

  @impl true
  def handle_in("signal", payload, socket) do
    broadcast_from!(socket, "signal", payload)
    {:noreply, socket}
  end
end

File: assets/js/features/workspace/services/signaling/signalingChannel.ts
import { Channel } from "phoenix";
import { connectSocket, socket } from "./socket";
import { createLogger } from "@/shared/lib/logger";

const log = createLogger("Signaling");

export class SignalingChannel {
    private channel: Channel;

    constructor(workspaceId: string) {
        this.channel = socket.channel(`signaling:${workspaceId}`);
    }

    join(): this {
        connectSocket();

        this.channel
            .join()
            .receive("ok", () => log.debug("joined"))
            .receive("error", (err) => log.error("join failed", err));
        return this;
    }

    leave(): void {
        this.channel.leave();
    }

    push(payload: object): void {
        this.channel.push("signal", payload);
    }

    on(callback: (payload: unknown) => void): number {
        return this.channel.on("signal", callback);
    }

    off(ref: number): void {
        this.channel.off("signal", ref);
    }
}
